### ISSUE TITLE
Cannot build on MacOS due to `fatal error: 'config.h' file not found`

### ISSUE THREAD

**Issue by joedborg** [cfallin/rust-oping#10]
This builds fine on my Ubuntu box, but fails as below on MacOS.  I've installed `liboping` via `brew`.

```
Compiling oping v0.3.3
error: failed to run custom build command for `oping v0.3.3`
process didn't exit successfully: `/Users/joe/Projects/ripping/target/release/build/oping-ecc6de6541c7cae2/build-script-build` (exit code: 101)
--- stdout
checking for a BSD-compatible install... /usr/local/bin/ginstall -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/local/bin/gmkdir -p
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking whether make supports the include directive... yes (GNU style)
checking dependency style of gcc... none
checking how to run the C preprocessor... gcc -E
checking whether ln -s works... yes
checking whether make sets $(MAKE)... (cached) yes
TARGET = Some("x86_64-apple-darwin")
OPT_LEVEL = Some("3")
TARGET = Some("x86_64-apple-darwin")
HOST = Some("x86_64-apple-darwin")
TARGET = Some("x86_64-apple-darwin")
TARGET = Some("x86_64-apple-darwin")
HOST = Some("x86_64-apple-darwin")
CC_x86_64-apple-darwin = None
CC_x86_64_apple_darwin = None
HOST_CC = None
CC = None
HOST = Some("x86_64-apple-darwin")
TARGET = Some("x86_64-apple-darwin")
HOST = Some("x86_64-apple-darwin")
CFLAGS_x86_64-apple-darwin = None
CFLAGS_x86_64_apple_darwin = None
HOST_CFLAGS = None
CFLAGS = None
DEBUG = Some("false")
running: "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-I" "liboping/src/" "-DHAVE_CONFIG_H" "-Wall" "-Wextra" "-o" "/Users/joe/Projects/ripping/target/release/build/oping-1e90dfaaa3ab24b9/out/liboping/src/liboping.o" "-c" "liboping/src/liboping.c"
cargo:warning=liboping/src/liboping.c:25:11: fatal error: 'config.h' file not found
cargo:warning=# include <config.h>
cargo:warning=          ^~~~~~~~~~
cargo:warning=1 error generated.
exit code: 1

--- stderr
aclocal: warning: couldn't open directory 'm4': No such file or directory
configure.ac:57: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:57: You should run autoupdate.
../../lib/autoconf/general.m4:207: AC_HELP_STRING is expanded from...
configure.ac:57: the top level
configure.ac:57: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:57: You should run autoupdate.
../../lib/autoconf/general.m4:207: AC_HELP_STRING is expanded from...
configure.ac:57: the top level
src/Makefile.am:10: error: Libtool library used but 'LIBTOOL' is undefined
src/Makefile.am:10:   The usual way to define 'LIBTOOL' is to add 'LT_INIT'
src/Makefile.am:10:   to 'configure.ac' and run 'aclocal' and 'autoconf' again.
src/Makefile.am:10:   If 'LT_INIT' is in 'configure.ac', make sure
src/Makefile.am:10:   its definition is in aclocal's search path.
autoreconf: automake failed with exit status: 1
./configure: line 4183: syntax error near unexpected token `dlopen'
./configure: line 4183: `LT_INIT(dlopen)'
thread 'main' panicked at '

Internal error occurred: Command "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-I" "liboping/src/" "-DHAVE_CONFIG_H" "-Wall" "-Wextra" "-o" "/Users/joe/Projects/ripping/target/release/build/oping-1e90dfaaa3ab24b9/out/liboping/src/liboping.o" "-c" "liboping/src/liboping.c" with args "cc" did not execute successfully (status code exit code: 1).

', /Users/joe/.cargo/registry/src/github.com-1ecc6299db9ec823/gcc-0.3.55/src/lib.rs:1672:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by cfallin** [comment]
Hmm. Thanks for this report. I don't have an OS X machine so I can't debug this any further, but I'd welcome a PR if you can work out the fix.

FWIW, the latest version of this crate actually bundles `liboping` directly and compiles it with a `build.rs` script. So installing it system-wide beforehand shouldn't be necessary. But perhaps Homebrew applies some patches first?

Looking at the error messages, my educated guesses are (i) something's slightly awry with the compiler invocation in build.rs (it explicitly undefines `HAVE_CONFIG_H`, so `liboping.c` shouldn't try to include it), or (ii) something's wrong with autotools on your install, perhaps old versions (see error in `Makefile.am`)?

Again, happy to take a PR if you can work this out!

**Comment by joedborg** [comment]
Thanks for the reply @cfallin, I'd completely overlooked the `build.rs` file.  I ran through that step by step and can see that `libtool` needed installing.  With Mac, thats via `brew install libtool` and Ubuntu `sudo apt-get install libtool`.  Once installing this on MacOS, I can build via Cargo.

**Comment by cfallin** [comment]
Great, I'm glad this was resolved! I added a note to `README.md` to list the build deps including `libtool` explicitly -- sorry about the lack of clarity there.

**Comment by joedborg** [comment]
@cfallin no problem at all!  thanks for adding it to the `README`.

**Comment by kolapapa** [comment]
@cfallin help~~

os version:
```shell
root@822d18fc18bc:~/test/hello# cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.2 LTS"
```
I installed all the packages on `README.md`.
such as `libtool`, `autoconf ` and `automake`
```shell
root@822d18fc18bc:~/test/hello# cargo build
   Compiling oping v0.3.3
error: failed to run custom build command for `oping v0.3.3`

Caused by:
  process didn't exit successfully: `/root/test/hello/target/debug/build/oping-ee35fb8a8a0336ec/build-script-build` (exit code: 101)
--- stdout
autoconfiguration done, to build: ./configure ; make
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of gcc... none
checking how to run the C preprocessor... gcc -E
checking whether ln -s works... yes
checking whether make sets $(MAKE)... (cached) yes
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking how to print strings... printf
checking for a sed that does not truncate output... /bin/sed
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for fgrep... /bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking the maximum length of command line arguments... 1572864
checking how to convert x86_64-pc-linux-gnu file names to x86_64-pc-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /bin/dd
checking how to truncate binary pipes... /bin/dd bs=4096 count=1
checking for mt... no
checking if : is a manifest tool... no
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking for shl_load... no
checking for shl_load in -ldld... no
checking for dlopen... no
checking for dlopen in -ldl... yes
checking whether a program can dlopen itself... yes
checking whether a statically linked program can dlopen itself... no
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for ANSI C header files... (cached) yes
checking whether time.h and sys/time.h may both be included... yes
checking math.h usability... yes
checking math.h presence... yes
checking for math.h... yes
checking signal.h usability... yes
checking signal.h presence... yes
checking for signal.h... yes
checking fcntl.h usability... yes
checking fcntl.h presence... yes
checking for fcntl.h... yes
checking for inttypes.h... (cached) yes
checking netdb.h usability... yes
checking netdb.h presence... yes
checking for netdb.h... yes
checking for stdint.h... (cached) yes
checking for stdlib.h... (cached) yes
checking for string.h... (cached) yes
checking sys/socket.h usability... yes
checking sys/socket.h presence... yes
checking for sys/socket.h... yes
checking sys/time.h usability... yes
checking sys/time.h presence... yes
checking for sys/time.h... yes
checking for unistd.h... (cached) yes
checking locale.h usability... yes
checking locale.h presence... yes
checking for locale.h... yes
checking langinfo.h usability... yes
checking langinfo.h presence... yes
checking for langinfo.h... yes
checking for netinet/in_systm.h... yes
checking for netinet/in.h... yes
checking for netinet/ip.h... yes
checking for netinet/ip_icmp.h... yes
checking for netinet/ip_var.h... no
checking for netinet/ip6.h... yes
checking for netinet/icmp6.h... yes
checking for size_t... yes
checking for ssize_t... yes
checking for uid_t in sys/types.h... yes
checking for uint16_t... yes
checking for uint32_t... yes
checking for uint8_t... yes
checking for socket... yes
checking for __xnet_recvmsg in -lxnet... no
checking for library containing nanosleep... none required
TARGET = Some("x86_64-unknown-linux-gnu")
OPT_LEVEL = Some("0")
TARGET = Some("x86_64-unknown-linux-gnu")
HOST = Some("x86_64-unknown-linux-gnu")
TARGET = Some("x86_64-unknown-linux-gnu")
TARGET = Some("x86_64-unknown-linux-gnu")
HOST = Some("x86_64-unknown-linux-gnu")
CC_x86_64-unknown-linux-gnu = None
CC_x86_64_unknown_linux_gnu = None
HOST_CC = None
CC = None
HOST = Some("x86_64-unknown-linux-gnu")
TARGET = Some("x86_64-unknown-linux-gnu")
HOST = Some("x86_64-unknown-linux-gnu")
CFLAGS_x86_64-unknown-linux-gnu = None
CFLAGS_x86_64_unknown_linux_gnu = None
HOST_CFLAGS = None
CFLAGS = None
DEBUG = Some("true")
running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-m64" "-I" "liboping/src/" "-DHAVE_CONFIG_H" "-Wall" "-Wextra" "-o" "/root/test/hello/target/debug/build/oping-c0f857d520f239df/out/liboping/src/liboping.o" "-c" "liboping/src/liboping.c"
cargo:warning=liboping/src/liboping.c:25:11: fatal error: config.h: No such file or directory
cargo:warning= # include <config.h>
cargo:warning=           ^~~~~~~~~~
cargo:warning=compilation terminated.
exit code: 1

--- stderr
configure.ac:57: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:57: You should run autoupdate.
../../lib/autoconf/general.m4:207: AC_HELP_STRING is expanded from...
configure.ac:57: the top level
configure.ac:57: warning: The macro `AC_HELP_STRING' is obsolete.
configure.ac:57: You should run autoupdate.
../../lib/autoconf/general.m4:207: AC_HELP_STRING is expanded from...
configure.ac:57: the top level
./configure: line 12854: syntax error near unexpected token `NCURSES,'
./configure: line 12854: `	PKG_CHECK_MODULES(NCURSES, ncursesw, can_build_with_ncurses=yes,'
thread 'main' panicked at '

Internal error occurred: Command "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-m64" "-I" "liboping/src/" "-DHAVE_CONFIG_H" "-Wall" "-Wextra" "-o" "/root/test/hello/target/debug/build/oping-c0f857d520f239df/out/liboping/src/liboping.o" "-c" "liboping/src/liboping.c" with args "cc" did not execute successfully (status code exit code: 1).

', /root/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/gcc-0.3.55/src/lib.rs:1672:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

```

Thanks.

**Comment by kolapapa** [comment]
```shell
sudo apt-get install libconfig-dev
```
I installed it will fix this problem.

**Comment by cfallin** [comment]
Hmm -- `liboping` seems to be having issues with its configure script. It's a separate package that is not part of this repository, but instead, imported as a submodule and bundled with the crate.

`config.h` should be generated by the configure script, but isn't, because the configure script has a syntax error (see the very end of your error output). `libconfig-dev` is a completely unrelated package and not the right solution. Could you check whether `liboping` by itself (https://github.com/octo/liboping) builds on your system?

**Comment by kolapapa** [comment]
@cfallin  Thank you for your help~~

 I seem to have installed this:
```shell
$ apt-get install liboping-dev oping liboping0
```

**Comment by nhed** [comment]
```configure.ac:57: You should run autoupdate.```
~@cfallin - I think u can solve in your `build.rs`, also autoconf is a pain-in-the-neck~

https://github.com/cfallin/rust-oping/issues/8#issuecomment-632996289

**Comment by cfallin** [comment]
@nhed possibly, though I'm trying not to spend too much time burying myself in autoconf hell these days :-) If there's a PR that would help (e.g. running autoupdate first -- I haven't done this yet as I don't have the ability to test this on non-Linux systems), I'm happy to review/merge. Thanks!

**Comment by nhed** [comment]
@cfallin you are right to avoid it :)

actually the fix you just picked from my colleague (in 0.3.5) should help folks avoid this if they can go the dynamic linking route.

The thing I didn't like about my own proposal is that it modified a source-controlled file in the sub-module :(       (and do now i'm gonna strikeout my orig comment)

**Comment by cfallin** [comment]
Just published 0.4.0, which removes the bundled `liboping` and requires it to be installed on the system instead, due to continued support burden of build issues (sorry!). Closing this issue now as it's no longer applicable.
